src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Helpers.cs

```diff
@@ -89,34 +89,82 @@ public static X509Certificate2 TryFindMatchingCertificate(this X509Certificate2C
                 case SubjectIdentifierType.IssuerAndSerialNumber:
                     {
                         X509IssuerSerial issuerSerial = (X509IssuerSerial)(recipientIdentifier.Value);
-
-                        using (ClonedCertificates matches1 = certs.FindCertificates(X509FindType.FindBySerialNumber, issuerSerial.SerialNumber))
+                        string serialNumber = issuerSerial.SerialNumber;
+                        string issuer = issuerSerial.IssuerName;
+                        foreach (X509Certificate2 candidate in certs)
                         {
-                            using (ClonedCertificates matches2 = matches1.FindCertificates(X509FindType.FindByIssuerDistinguishedName, issuerSerial.IssuerName))
-                            {
-                                // Desktop compat: We do not complain about multiple matches. Just take the first one and ignore the rest.
-                                return matches2.Any ? matches2.ClaimResult() : null;
-                            }
+                            if (candidate.SerialNumber == serialNumber && candidate.Issuer == issuer)
+                                return candidate;
                         }
                     }
+                    break;
 
                 case SubjectIdentifierType.SubjectKeyIdentifier:
                     {
-                        string ski = (string)(recipientIdentifier.Value);
-                        using (ClonedCertificates matches = certs.FindCertificates(X509FindType.FindBySubjectKeyIdentifier, ski))
+                        string skiString = (string)(recipientIdentifier.Value);
+                        byte[] ski = skiString.ToSki();
+                        foreach (X509Certificate2 cert in certs)
                         {
-                            // Desktop compat: We do not complain about multiple matches. Just take the first one and ignore the rest.
-                            return matches.Any ? matches.ClaimResult() : null;
+                            byte[] candidateSki = PkcsPal.Instance.GetSubjectKeyIdentifier(cert);
```

Comment:
Hmm, actually the extensions would be a potential breaking change.  The CERT_KEY_IDENTIFIER_PROP_ID property is documented as making up an answer when the extension isn't present.